p2p/protocols/protocol.go

```diff
@@ -288,40 +384,46 @@ func (p *Peer) Send(ctx context.Context, msg interface{}) error {
 	return p2p.Send(p.rw, code, wmsg)
 }
 
-// handleIncoming(code)
-// is called each cycle of the main forever loop that dispatches incoming messages
-// if this returns an error the loop returns and the peer is disconnected with the error
-// this generic handler
-// * checks message size,
-// * checks for out-of-range message codes,
-// * handles decoding with reflection,
-// * call handlers as callbacks
-func (p *Peer) handleIncoming(handle func(ctx context.Context, msg interface{}) error) error {
+// Receive(code) is a sync call that handles incoming message with provided message handler
+func (p *Peer) Receive(handler func(ctx context.Context, msg interface{}) error) error {
 	msg, err := p.rw.ReadMsg()
 	if err != nil {
 		return err
 	}
-	// make sure that the payload has been fully consumed
+
+	return p.handleMsg(&msg, handler)
+}
+
+// handleMsg is handling message with provided handler. It:
+// * checks message size,
+// * checks for out-of-range message codes,
+// * handles decoding with reflection,
+// * call handlers as callbacks
+func (p *Peer) handleMsg(msg *p2p.Msg, handle func(ctx context.Context, msg interface{}) error) error {
+	// make sure that the payload has been fully consume
 	defer msg.Discard()
 
 	if msg.Size > p.spec.MaxMsgSize {
-		return errorf(ErrMsgTooLong, "%v > %v", msg.Size, p.spec.MaxMsgSize)
+		err := errorf(ErrMsgTooLong, "%v > %v", msg.Size, p.spec.MaxMsgSize)
```

Comment:
why these changes needed?

Reply:
Oh, sorry, it's probably just my debug leftover 🙈